src/MongoDB.Driver/Linq/FieldExpressionFlattener.cs

```diff
@@ -86,5 +80,43 @@ protected internal override Expression VisitField(FieldExpression node)
 
             return node;
         }
+
+        // private methods
+        private static string FormatArrayIndex(object index)
+        {
+            if (index is int intIndex)
+            {
+                return FormatArrayIndex((long)intIndex);
+            }
+
+            if (index is long longIndex)
+            {
+                return FormatArrayIndex(longIndex);
+            }
+
+            throw new NotSupportedException("Array indexes must be int or long.");
+        }
+
+        private static string FormatArrayIndex(long index) 
+        {
+            switch (index)
+            {
+                // We've treated -1 as meaning $ operator. We can't break this now,
+                // so, specifically when we are flattening fields names, this is 
+                // how we'll continue to treat -1.
+                case var _ when index < -1L:
+                {
+                    throw new NotSupportedException("Array indexes must be more or equal than -1.");
+                }
+                case var _ when index == -1L:
+                {
+                    return "$";
+                }
+                default:
+                {
+                    return index.ToString(NumberFormatInfo.InvariantInfo);
+                }
```

Comment:
The `{` and `}` are not needed on every case are they?

Reply:
Sure, done

Reply:
They are still there.